ksql-engine/src/main/java/io/confluent/ksql/topic/DefaultTopicInjector.java

```diff
@@ -0,0 +1,202 @@
+            cas.getStatement().getQuery(),
+            Optional.of(cas.getStatement().getSink()));
+
+    final StructuredDataSource theSource = analysis.getTheSource();
+    final String kafkaTopicName = theSource.getKsqlTopic().getKafkaTopicName();
+    return topicClient.describeTopic(kafkaTopicName);
+  }
+
+  private static String topicName(
+      final CreateAsSelect cas,
+      final KsqlConfig cfg
+  ) {
+    final Expression topicProperty = cas.getProperties().get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY);
+    return (topicProperty == null)
+        ? cfg.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG) + cas.getName().getSuffix()
+        : StringUtils.strip(topicProperty.toString(), "'");
+  }
+
+  private static int numPartitions(
+      final PreparedStatement<? extends CreateAsSelect> cas,
+      final KsqlConfig cfg,
+      final Map<String, Object> propertyOverrides,
+      final TopicDescription description
+  ) {
+    final Optional<Integer> inProperties = checkedParse(
+        Integer::parseInt, cas, KsqlConstants.SINK_NUMBER_OF_PARTITIONS);
+    if (inProperties.isPresent()) {
+      return inProperties.get();
+    }
+
+    final Integer legacyNumPartitions = cfg.getInt(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY);
+    return checkedParse(
+          Integer::parseInt,
+          KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY,
+          propertyOverrides)
+        .orElseGet(() -> ObjectUtils.defaultIfNull(
+            legacyNumPartitions,
+            description.partitions().size()));
+  }
+
+  private static short numReplicas(
+      final PreparedStatement<? extends CreateAsSelect> cas,
+      final KsqlConfig cfg,
+      final Map<String, Object> propertyOverrides,
+      final TopicDescription description
+  ) {
+    final Optional<Short> inProperties = checkedParse(
+        Short::parseShort, cas, KsqlConstants.SINK_NUMBER_OF_REPLICAS);
+    if (inProperties.isPresent()) {
+      return inProperties.get();
+    }
+
+    final Short legacyNumReplicas = cfg.getShort(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY);
+    return checkedParse(
+          Short::parseShort,
+          KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY,
+          propertyOverrides)
+        .orElseGet(() -> ObjectUtils.defaultIfNull(
+            legacyNumReplicas,
+            (short) description.partitions().get(0).replicas().size()));
+  }
+
+  private static <T extends Number> Optional<T> checkedParse(
+      final Function<String, T> parse,
+      final PreparedStatement<? extends CreateAsSelect> cas,
+      final String property
+  ) {
+    final Object exp = cas.getStatement().getProperties().get(property);
+    if (exp == null) {
+      return Optional.empty();
+    }
+
+    final String val = exp.toString();
+    try {
+      return Optional.ofNullable(parse.apply(val));
+    } catch (NumberFormatException e) {
+      throw new KsqlStatementException(
+          String.format("Invalid value for property %s: %s",
```

Comment:
The error message does not explaining _why_ the value is invalid, and you're also not include the original exception `e` in the new exception.

The unit tests you've added actually rely on the code in `Analyzer` to validate these properties, not this code.